jerry-core/ecma/builtin-objects/ecma-builtin-date.cpp

```diff
@@ -38,19 +41,258 @@
+ */
+static ecma_number_t
+ecma_date_parse_date_chars (lit_utf8_byte_t *date_start_p, /**< start pointer of the utf8 string */
+                            lit_utf8_byte_t **date_char_p, /**< current pointer of the utf8 string */
+                            lit_utf8_size_t date_str_size, /**< size of the utf8 string */
+                            uint32_t num_of_chars) /**< number of characters to read and convert */
+{
+  ecma_number_t ret_value = ecma_number_make_nan ();
+  JERRY_ASSERT ((*date_char_p - date_start_p) >= 0);
+  uint32_t num_of_visited_chars = (uint32_t) (*date_char_p - date_start_p);
+
+  if ((date_str_size - num_of_visited_chars) >= num_of_chars)
+  {
+    ecma_string_t *str_p = ecma_new_ecma_string_from_utf8 (*date_char_p, num_of_chars);
+    ret_value = ecma_string_to_number (str_p);
+    ecma_deref_ecma_string (str_p);
+
+    *date_char_p += num_of_chars * sizeof (lit_utf8_byte_t);
+  }
+
+  return ret_value;
+} /* ecma_date_parse_date_chars */
+
+/**
  * The Date object's 'parse' routine
  *
  * See also:
  *          ECMA-262 v5, 15.9.4.2
+ *          ECMA-262 v5, 15.9.1.15
  *
  * @return completion value
  *         Returned value must be freed with ecma_free_completion_value.
  */
 static ecma_completion_value_t
-ecma_builtin_date_parse (ecma_value_t this_arg, /**< this argument */
+ecma_builtin_date_parse (ecma_value_t this_arg __attr_unused___, /**< this argument */
                          ecma_value_t arg) /**< string */
 {
-  ECMA_BUILTIN_CP_UNIMPLEMENTED (this_arg, arg);
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+  ecma_number_t *date_num_p = ecma_alloc_number ();
+  *date_num_p = ecma_number_make_nan ();
+  ecma_number_t time = ECMA_NUMBER_ZERO;
+
+  /* Date Time String fromat (ECMA-262 v5, 15.9.1.15) */
+  ecma_string_t *date_str_p;
+  ECMA_TRY_CATCH (date_str_value,
+                  ecma_op_to_string (arg),
+                  ret_value);
+
+  date_str_p = ecma_get_string_from_value (date_str_value);
+
+  lit_utf8_size_t date_str_size = ecma_string_get_size (date_str_p);
+  MEM_DEFINE_LOCAL_ARRAY (date_start_p, date_str_size, lit_utf8_byte_t);
+
+  ecma_string_to_utf8_string (date_str_p, date_start_p, (ssize_t) date_str_size);
+
+  lit_utf8_byte_t *date_char_p = date_start_p;
+
+  ecma_number_t year;
+  ecma_number_t month = ECMA_NUMBER_ONE;
+  ecma_number_t day = ECMA_NUMBER_ONE;
+  ecma_number_t hours = ECMA_NUMBER_ZERO;
+  ecma_number_t minutes = ECMA_NUMBER_ZERO;
+  ecma_number_t seconds = ECMA_NUMBER_ZERO;
+  ecma_number_t milliseconds = ECMA_NUMBER_ZERO;
+
+  /* 1. read year */
+  year = ecma_date_parse_date_chars (date_start_p, &date_char_p, date_str_size, 4);
+
+  if (ecma_number_is_nan (year)
+      || year < 0)
+  {
+    goto end; /* NaN */
+  }
+
+  /* 2. read month if any */
+  if (date_str_size != (lit_utf8_size_t) (date_char_p - date_start_p)
+      && *date_char_p == '-')
+  {
+    /* eat up '-' */
+    date_char_p += sizeof (lit_utf8_byte_t);
```

Comment:
Btw, is it correct way to iterate string? Can't the string contain a character, encoded with a multi-byte sequence?